lib/cc/engine/analyzers/issue.rb

```diff
@@ -0,0 +1,89 @@
+require "cc/engine/analyzers/sexp_lines"
+require "cc/engine/analyzers/violation_read_up"
+require "digest"
+
+module CC
+  module Engine
+    module Analyzers
+      class Issue
+        def initialize(language_strategy:, check_name:, current_sexp:, other_sexps:)
+          @language_strategy = language_strategy
+          @check_name = check_name
+          @current_sexp = current_sexp
+          @other_sexps = other_sexps
+        end
+
+        def format # rubocop:disable Metrics/MethodLength
+          {
+            "type": "issue",
+            "check_name": check_name,
+            "description": description,
+            "categories": ["Duplication"],
+            "location": format_location,
+            "remediation_points": calculate_points,
+            "other_locations": format_other_locations,
+            "content": content_body,
+            "fingerprint": fingerprint,
+          }
+        end # rubocop:enable Metrics/MethodLength
+
+        def report_name
+          "#{current_sexp.file}-#{current_sexp.line}"
+        end
+
+        def mass
+          current_sexp.mass
+        end
+
+        private
+
+        attr_reader :language_strategy, :check_name, :other_sexps, :current_sexp
+
+        def calculate_points
+          language_strategy.calculate_points(mass)
+        end
+
+        def format_location
+          format_sexp(current_sexp)
+        end
+
+        def format_other_locations
+          other_sexps.map do |sexp|
+            format_sexp(sexp)
+          end
+        end
+
+        def format_sexp(sexp)
+          lines = SexpLines.new(sexp)
+          {
+            "path": sexp.file.gsub(%r{^./}, ""),
+            "lines": {
+              "begin": lines.begin_line,
+              "end": lines.end_line,
+            },
+          }
+        end
+
+        def content_body
+          { "body": ViolationReadUp.new(mass).contents }
+        end
+
+        def fingerprint
+          digest = Digest::MD5.new
+          digest << current_sexp.file
```

Comment:
We did a lot of work semi-recently to fix comparison differences between app/compare & finalizer. And when we shipped the new PR header, it was considered critical that the app/compare's pass/fail logic match finalizer's. I think this fingerprinting algorithm will break those expectations.

Example:

in the master snapshot, file `foo.rb` has the following contents:

```
class A
  def foo
     # some large duplicated code block here
  end

  def bar
     # the same code block here
  end
end
```

That will result in 2 issues with the same fingerprint, but it's the master snapshot, so who cares.

In the pull request snapshot, file `foo.rb` looks like this:

```
class A
  def foo
     # some large duplicated code block here
  end

  def bar
     # the same code block here
  end

  def baz
    # the same code block here, again
  end
end
```

Now we have 3 issues on `foo.rb`, all with the same snapshot. (And the newest issue for `baz` doesn't even have to be same code: as I said, it could happen to be totally different code duplicated elsewhere which happens to have the same mass as `foo` and `bar`.)

Finalizer is going to see no new fingerprints, and pass the PR. If the user goes to the PR compare page, I *think* they'll see the "passed" header, but in the issues area they'll see 1 new issue.